celltype_annotation/analysis/copykat-rms-exploration.Rmd

```diff
@@ -0,0 +1,224 @@
+                         "analysis", 
+                         "copykat-results")
+fs::dir_create(output_dir)
+
+copykat_sce <- file.path(output_dir,
+                         glue::glue("{params$library}_copykat_sce.rds"))
+```
+
+In the next chunk, we obtain the _submitter-annotated_ SCE from S3, if it is not already locally present.
+We'll wait to read it in when actually running `copyKat`, mostly for memory reasons.
+
+
+```{r}
+# make local data directory if it doesn't exist
+if (!dir.exists(params$local_data_dir)) {
+  dir.create(params$local_data_dir, recursive = TRUE)
+}
+
+# build path to submitter annotated sce file 
+submitter_annotated_sce_file <- glue::glue("{params$library_id}_processed_celltype.rds")
+
+
+local_submitter_sce_path <- file.path(params$local_data_dir,
+                                      params$sample_id,
+                                      submitter_annotated_sce_file)
+
+# get submitter sce from S3 if not already present locally
+if (!(file.exists(local_submitter_sce_path))) {
+  sync_call <- glue::glue('op run -- aws s3 sync {params$s3_submitter_data_dir} {params$local_data_dir}/{params$sample_id} --exclude "*" --include "{submitter_annotated_sce_file}"')
+  system(sync_call, ignore.stdout = TRUE) 
+}
+```
+
+
+## Run CopyKAT
+
+Now, we are ready to run `copyKat` on the SCE, using the _raw expression matrix_ as described in the [GitHub `README` file](https://github.com/navinlabcode/copykat/blob/master/README.md).
+
+Of note, the `README` includes this caveat about the `"distance"` parameter:
+
+> One struggling observation is that none of one clustering method could fit all datasets. In this version, I add a distance parameters for clustering that include "euclidean" distance and correlational distance, ie. 1-"pearson" and "spearman" similarity. In general, corretional distances tend to favor noisy data, while euclidean distance tends to favor data with larger CN segments.
+
+Therefore we'll compare `"euclidian"`, `"spearman"`, and `"pearson"` values for [this parameter](https://github.com/navinlabcode/copykat/blob/b795ff793522499f814f6ae282aad1aab790902f/R/copykat.R#L251-L255).
+
+
+```{r message=FALSE, warning=FALSE}
+# Only run copykat if the final SCE does not exist
+if (file.exists(copykat_sce)) {
+  
+  sce <- readr::read_rds(copykat_sce)
+  
+} else {
+  
+  # Otherwise, read in initial SCE and run copyKat
+  sce <- readr::read_rds(local_submitter_sce_path)
+    
+  purrr::walk(
+    c("spearman", "pearson", "euclidean"), 
+    \(x) {
+      
+      copykat_result <- copykat(
+        rawmat = as.matrix(counts(sce)), 
+        id.type = "E", # we have Ensembl gene ids, not the default gene symbols ("S")
+        sam.name = params$library_id, 
+        distance = x, 
+        plot.genes = FALSE,
+        n.cores = params$threads
+      )
+      
+      # save prediction to sce
+      colname <- glue::glue("copykat_{x}")
+      colData(sce)[[colname]] <- copykat_result$prediction$copykat.pred
```

Comment:
What else is output by copykat? Anything that might be worth saving besides the predictions? The contents of the [`test_output` folder on their repo](https://github.com/navinlabcode/copykat/tree/master/test_output) has a `_CNA_results.txt` matrix? Maybe that's helpful in figuring out confidence or scores?

Reply:
All of that output is contained within the `copyKat` object itself that I saved (but did not commit!), including that matrix. It doesn't appear to be terribly useful, though. From the README - 

> The first 3 columns in the CNV matrix are the genomic coordinates. Rows are 220KB bins in genomic orders. It also output CNVs indexed by gene names and other information such as chromosome names, start and end positions, G staining band etc.

In addition to those first 3 columns, it then has 1 column _per cell_ with information for each 220KB bin. Here's a bit of what it looks like, just showing 2 cells. It appears to mostly (only?) be useful for making the heatmap, which doesn't relate to our needs. 

```
> result$CNAmat[1:20, 1:5]
   chrom chrompos  abspos GACTCTCCACAAGCTT TTGCCTGGTTTACTTC
1      1  1042457 1042457      -0.06569387       0.02518337
2      1  1265484 1265484      -0.06569387       0.02518337
3      1  1519859 1519859      -0.06569387       0.02518337
4      1  1826619 1826619      -0.06569387       0.02518337
5      1  2058465 2058465      -0.06569387       0.02518337
6      1  2280372 2280372      -0.06569387       0.02518337
7      1  2491263 2491263      -0.06569387       0.02518337
8      1  2817807 2817807      -0.06569387       0.02518337
9      1  3025418 3025418      -0.06569387       0.02518337
10     1  3230151 3230151      -0.06569387       0.02518337
11     1  3437000 3437000       0.05664600      -0.10335066
12     1  3645603 3645603       0.05664600      -0.10335066
13     1  4091348 4091348       0.05664600      -0.10335066
14     1  4312450 4312450       0.05664600      -0.10335066
15     1  4522391 4522391       0.05664600      -0.10335066
16     1  4733807 4733807       0.05664600      -0.10335066
17     1  4949751 4949751       0.05664600      -0.10335066
18     1  5167781 5167781       0.05664600      -0.10335066
19     1  5378577 5378577       0.05664600      -0.10335066
20     1  5586352 5586352       0.05664600      -0.10335066
```